libsignal-service-actix/src/push_service.rs

```diff
@@ -26,30 +27,36 @@ impl PushService for AwcPushService {
     type ByteStream = Box<dyn futures::io::AsyncRead + Unpin>;
     type WebSocket = AwcWebSocket;
 
-    async fn get<T>(&mut self, path: &str) -> Result<T, ServiceError>
+    async fn get_json<T>(
+        &mut self,
+        endpoint: Endpoint,
+        path: &str,
+        credentials: Option<(String, String)>,
+    ) -> Result<T, ServiceError>
     where
         for<'de> T: Deserialize<'de>,
     {
-        // In principle, we should be using http::uri::Uri,
-        // but that doesn't seem like an owned type where we can do this kind of
-        // constructions on.
-        // https://docs.rs/http/0.2.1/http/uri/struct.Uri.html
-        let url = self.base_url.join(path).expect("valid url");
+        let url = self.cfg.base_url(endpoint).join(path)?;
 
         log::debug!("AwcPushService::get({:?})", url);
         use awc::error::{ConnectError, SendRequestError};
-        let mut response = self.client.get(url.as_str()).send().await.map_err(
-            |e| match e {
-                SendRequestError::Connect(ConnectError::Timeout) => {
-                    ServiceError::Timeout {
-                        reason: e.to_string(),
-                    }
-                }
-                _ => ServiceError::SendError {
+        let mut request = self.client.get(url.as_str());
+        if let Some((ident, pass)) =
+            credentials.as_ref().or(self.credentials.as_ref())
+        {
+            request = request.basic_auth(ident, Some(pass));
+        }
+
+        let mut response = request.send().await.map_err(|e| match e {
+            SendRequestError::Connect(ConnectError::Timeout) => {
+                ServiceError::Timeout {
                     reason: e.to_string(),
-                },
+                }
+            }
+            _ => ServiceError::SendError {
+                reason: e.to_string(),
             },
-        )?;
+        })?;
```

Comment:
This code duplication between all the get/post/put methods should probably be macro'd away. That's probably for #63. That should also get rid of your concern for adding the authorization header manually...

Reply:
I ended up using `client::request` which feels much better now!